api/attachments_api_test.py

```diff
@@ -0,0 +1,175 @@
+from api import attachments_api
+from internals.core_enums import *
+from internals.core_models import FeatureEntry
+from internals import attachments
+
+test_app = flask.Flask(__name__)
+test_app.secret_key ='test'
+
+
+class AttachmentsAPITest(testing_config.CustomTestCase):
+
+  def setUp(self):
+    self.feature = FeatureEntry(
+        name='feat', summary='sum', category=1,
+        owner_emails=['owner@chromium.org'],
+        impl_status_chrome=ENABLED_BY_DEFAULT)
+    self.feature.put()
+
+    self.feature_id = self.feature.key.integer_id()
+    self.request_path = f'/api/v0/features/{self.feature_id}/attachments'
+    self.handler = attachments_api.AttachmentsAPI()
+
+  def tearDown(self):
+    testing_config.sign_out()
+    kinds: list[ndb.Model] = [FeatureEntry, attachments.Attachment]
+    for kind in kinds:
+      for entity in kind.query():
+        entity.key.delete()
+
+  def test_do_post__anon(self):
+    """Anon users cannot add attachments."""
+    testing_config.sign_out()
+    with test_app.test_request_context(self.request_path):
+      with self.assertRaises(werkzeug.exceptions.Forbidden):
+        self.handler.do_post(feature_id=self.feature_id)
+
+  def test_do_post__unregistered(self):
+    """Users who cannot create features cannot add attachments."""
+    testing_config.sign_in('someone@example.com', 111)
+    with test_app.test_request_context(self.request_path):
+      with self.assertRaises(werkzeug.exceptions.Forbidden):
+        self.handler.do_post(feature_id=self.feature_id)
+
+  def test_do_post__noneditor(self):
+    """Users who cannot edit this particular feature cannot add attachments."""
+    testing_config.sign_in('someone@example.com', 111)
+    with test_app.test_request_context(self.request_path):
+      with self.assertRaises(werkzeug.exceptions.Forbidden):
+        self.handler.do_post(feature_id=self.feature_id)
+
+  def test_do_post__no_files(self):
+    """Reject requests that have no attachments."""
+    testing_config.sign_in('owner@chromium.org', 111)
+    with test_app.test_request_context(self.request_path):
+      with self.assertRaises(werkzeug.exceptions.BadRequest):
+        self.handler.do_post(feature_id=self.feature_id)
+
+  def test_do_post__empty_file(self):
+    """Reject requests where the user did not upload."""
+    testing_config.sign_in('owner@chromium.org', 111)
+    body = ''
+    with test_app.test_request_context(self.request_path, data=body):
+      with self.assertRaises(werkzeug.exceptions.BadRequest):
+        self.handler.do_post(feature_id=self.feature_id)
+
+  def test_do_post__valid_file(self):
+    """With a valid user and valid request, we store the attachment."""
+    testing_config.sign_in('owner@chromium.org', 111)
+    mock_files = {'uploaded-file': testing_config.Blank(
+        filename='hello_attach.txt',
+        read=lambda: b'hello attachments!',
+        mimetype='text/plain')}
+    with test_app.test_request_context(self.request_path):
```

Comment:
```suggestion
    mock_file = (io.BytesIO(b'hello attachments!'), 'test.txt')
    with test_app.test_request_context(path=self.request_path, data={'uploaded-file': mock_file}):
```

Reply:
Thanks for that suggestion.  I had spent half the day struggling with passing in a complete POST body as a binary string.